proxy/proxy.go

```diff
@@ -72,10 +75,64 @@ type Proxy struct {
 	weaveWaitNoopVolume string
 	normalisedAddrs     []string
 	waiters             map[*http.Request]*wait
+	attachJobs          map[string]*attachJob
+	quit                chan struct{}
+}
+
+type attachJob struct {
+	id          string
+	tryInterval time.Duration // retry delay on next failure
+	timer       *time.Timer
+}
+
+func (proxy *Proxy) attachWithRetry(id string) {
+	proxy.Lock()
+	defer proxy.Unlock()
+	if j, ok := proxy.attachJobs[id]; ok {
+		j.timer.Reset(time.Duration(0))
+		return
+	}
+
+	j := &attachJob{id: id, tryInterval: initialInterval}
+	j.timer = time.AfterFunc(time.Duration(0), func() {
+		if err := proxy.attach(id, false); err != nil {
+			j.reschedule()
+			return
+		}
+		proxy.notifyWaiters(id, nil)
+	})
+	proxy.attachJobs[id] = j
```

Comment:
Yeah, inlining it does make it easier to follow, thanks. One thing I noticed while doing that, is that on line 92, in the case where an attach already exists, it short-circuits the current "retry timer", but doesn't reset the interval back down to 2 seconds. I'm debating if it should...